File: vinyl-catalog/src/main/updater.ts
/**
 * Auto-updater using electron-updater (GitHub Releases provider).
 *
 * Behaviour:
 * - Skipped entirely in dev mode
 * - Silently downloads updates in the background
 * - Installs automatically on next app quit (autoInstallOnAppQuit)
 * - Checks on startup, then every 4 hours
 */

import { autoUpdater } from 'electron-updater'
import { is } from '@electron-toolkit/utils'

const CHECK_INTERVAL_MS = 4 * 60 * 60 * 1000 // 4 hours

export function initAutoUpdater(): void {
  if (is.dev) return

  autoUpdater.autoDownload = true
  autoUpdater.autoInstallOnAppQuit = true

  autoUpdater.on('checking-for-update', () => {
    console.log('[updater] Checking for updates…')
  })

  autoUpdater.on('update-available', (info) => {
    console.log(`[updater] Update available: v${info.version}`)
  })

  autoUpdater.on('update-not-available', () => {
    console.log('[updater] App is up to date')
  })

  autoUpdater.on('update-downloaded', (info) => {
    console.log(`[updater] v${info.version} downloaded — will install on quit`)
  })

  autoUpdater.on('error', (err) => {
    // Non-fatal: log and swallow — update failure should never crash the app
    console.error('[updater] Error:', err.message)
  })

  // Initial check shortly after startup, then on a recurring interval
  setTimeout(() => autoUpdater.checkForUpdates().catch(() => {}), 10_000)
  setInterval(() => autoUpdater.checkForUpdates().catch(() => {}), CHECK_INTERVAL_MS)
}

File: vinyl-catalog/src-tauri/src/updater.rs
/*!
 * Auto-updater — wraps tauri-plugin-updater.
 *
 * Behaviour:
 * - Skipped in debug builds
 * - Checks for update 10 s after startup, then every 4 hours
 * - Downloads and installs silently; restart required to apply
 */

use std::time::Duration;
use tauri::AppHandle;
use tauri_plugin_updater::UpdaterExt;

const CHECK_INTERVAL: Duration = Duration::from_secs(4 * 60 * 60);

pub fn start(app: AppHandle) {
    if cfg!(debug_assertions) {
        return;
    }
    tauri::async_runtime::spawn(async move {
        // Give the app a moment to finish startup
        tokio::time::sleep(Duration::from_secs(10)).await;
        check_once(&app).await;

        let mut interval = tokio::time::interval(CHECK_INTERVAL);
        interval.tick().await; // skip immediate first tick
        loop {
            interval.tick().await;
            check_once(&app).await;
        }
    });
}

async fn check_once(app: &AppHandle) {
    log::info!("[updater] checking for updates…");
    match app.updater() {
        Ok(updater) => match updater.check().await {
            Ok(Some(update)) => {
                log::info!("[updater] update available: v{}", update.version);
                if let Err(e) = update.download_and_install(|_, _| {}, || {}).await {
                    log::error!("[updater] install failed: {}", e);
                } else {
                    log::info!("[updater] update installed — restart to apply");
                }
            }
            Ok(None) => log::info!("[updater] up to date"),
            Err(e) => log::warn!("[updater] check failed: {}", e),
        },
        Err(e) => log::warn!("[updater] updater unavailable: {}", e),
    }
}
